src/vm/tieredcompilation.cpp

```diff
@@ -421,52 +498,90 @@ void TieredCompilationManager::OptimizeMethodsCallback()
         }
     }
 
-    ULONGLONG startTickCount = CLRGetTickCount64();
-    NativeCodeVersion nativeCodeVersion;
     EX_TRY
     {
         GCX_COOP();
         ENTER_DOMAIN_ID(m_domainId);
         {
-            GCX_PREEMP();
-            while (true)
+            OptimizeMethods();
+        }
+        END_DOMAIN_TRANSITION;
+    }
+    EX_CATCH
+    {
+        STRESS_LOG1(LF_TIEREDCOMPILATION, LL_ERROR, "TieredCompilationManager::OptimizeMethodsCallback: "
+            "Unhandled exception on domain transition, hr=0x%x\n",
+            GET_EXCEPTION()->GetHR());
+    }
+    EX_END_CATCH(RethrowTerminalExceptions);
+}
+
+//This method will process one or more methods from optimization queue
+// on a background thread. Each such method will be jitted with code
+// optimizations enabled and then installed as the active implementation
+// of the method entrypoint.
+// 
+// We need to be carefuly not to work for too long in a single invocation
+// of this method or we could starve the threadpool and force
+// it to create unnecessary additional threads.
+void TieredCompilationManager::OptimizeMethods()
+{
+    WRAPPER_NO_CONTRACT;
+
+    ULONGLONG startTickCount = CLRGetTickCount64();
+    NativeCodeVersion nativeCodeVersion;
+    EX_TRY
+    {
+        GCX_PREEMP();
+        while (true)
+        {
             {
+                SpinLockHolder holder(&m_lock); 
+                nativeCodeVersion = GetNextMethodToOptimize();
+                if (nativeCodeVersion.IsNull() ||
+                    m_isAppDomainShuttingDown)
                 {
-                    SpinLockHolder holder(&m_lock); 
-                    nativeCodeVersion = GetNextMethodToOptimize();
-                    if (nativeCodeVersion.IsNull() ||
-                        m_isAppDomainShuttingDown)
-                    {
-                        DecrementWorkerThreadCount();
-                        break;
-                    }
-                    
+                    DecrementWorkerThreadCount();
```

Comment:
yep you are right... let me see if that sinks it or there is a modified repro ; )

Reply:
> a) Using the timer callback thread optionally as a background method compilation thread introduces multiple code flow paths into the same async work. Either we should keep the timer callback fully separate, or clearly define the invariants that are shared across all worker threads and if possible use a shared code path to deal with shared invariants.

It just felt unnecessary to queue to the thread pool again when we already have a thread pool thread ready to optimize. The invariants to the root of the shared code path (OptimizeMethods) I think are:
- thread-running count is 1
- the thread has already entered the app domain

I should add asserts for those to OptimizeMethods to state the preconditions. Do you have other ideas?

> b) We've got two different locks protecting different pieces of state which makes it trickier to reason about the allowable states. I suspect we could converge to a single spin lock? For example m_methodsToOptimize is protected by spin lock and m_hasMethodsToOptimizeAfterDelay is protected under the Crst. If there was a single lock I think you could get rid m_hasMethodsToOptimizeAfterDelay and just check whether or not queued work exists.

I had to change the spin lock for the call counting delay into a crst because apparently you can't enter a crst from inside a spin lock. That lock protects the fields immediately following it in the .h file. Currently the two locks protect distinct things also such that they would not need to be nested. They could be combined for simplicity, I doubt it would make any difference since both locks are typically held for a short duration, but it would have to be a crst. Probably crst wouldn't make much difference from a spin lock either since it's unlikely to be contended for very long or too frequently.

Reply:
Another thought is maybe all of the call counting delay stuff can be separated out into a separate class

Reply:
But I think merging the locks would be fine for now

Reply:
I think I am getting better grasp on what the delayed queueing invariants are, I'm still thinking if I've got more precise suggestions on what to change or maybe its just a matter of comments to explain the invariants. I'll keep thinking on it but have a good vacation!

Reply:
By the end I felt fairly convinced what you had was correct, it just felt hard to reason about it or how it would be affected by further modifications. I messed around with some refactoring in my fork in the TierFix branch. Aside from just breaking down a few methods into smaller pieces I also merged the locks and eliminated m_hasMethodsToOptimizeAfterDelay in favor of being able to recalculate at any time if another worker thread is needed. I haven't tested it nor am I saying you should definitely do it that way, but I think its worth a look. I think the main things I liked refactoring this way:
a) single lock feels easier to reason about state changes
b) worker thread count again represents threads that are actively running (or queued to run imminently)
c) seems like its closer to what we would need if we wanted to increase parallelism or drive Pause/Resume with other triggering mechanisms.

Reply:
Thanks @noahfalk. I have folded some suggestions from your fork into the change:
- Merged locks as Crst, see comment above call to CreateTimerQueueTimer. Perf doesn't seem to be affected.
- Refactored thread count increment and queuing to thread pool into separate functions
- Eliminated m_hasMethodsToOptimizeAfterDelay and used your mechanism instead
- If we would need to add manual pause/resume capabilities, there may be more things to take care of
  - Such as:
    - Keeping track of nested manual pause requests such that tiering is not resumed until all pausers have requested to resume
    - Creating/deleting the timer at the appropriate times
    - Syncing manual resumes with the automatic resume from the timer
  - It would be possible to do if we need that capability, but it seems like it is adding complication by introducing issues that don't exist currently and may or may not exist in the future, so I have left that out

Reply:
Looks good to me!